### ISSUE TITLE
pam_mount and systemd-user - (pam_mount.c:477): warning: could not obtain password interactively

### ISSUE THREAD

**Issue by voobscout** [NixOS/nixpkgs#50447]
Potentially fixes #44586, #21314 and obsoletes dc80e99c7741b1bd82bf73827b4e920daa223e73

The problem is, currently pam_mount is being called twice, so it screws things up the 2nd time. It creates many unpleasant side effects, especially when mounting user homes with fuse (encfs/cryfs/gocryptofs).

Example /etc/static/security/pam_mount.conf.xml that doesn't work out of the box:
```xml
<volume user="test_cryfs" fstype="fuse" options="nodev,nosuid,nonempty,allow_other" path="cryfs#/home/cipher/test_cryfs" mountpoint="/home/test_cryfs" />
```
Symptoms include, but are not limited to - lack of dbus env vars when /run/user/UID is present and logs show systemd-user session being started, failure to umount user home upon logout, nixos-rebuild asking for pam_mount password twice.

Here is the culprit:
https://github.com/NixOS/nixpkgs/blob/db0abe98bfa7adad39912edad44743de030f8300/nixos/modules/security/pam.nix#L385-L388
Insert after L385 and before L386:
```
session   [success=1 default=ignore]  pam_succeed_if.so  service = systemd-user quiet
```
and problem solved.

Can someone please perma-fix this little snafu? Took me 3 days to figure it out. I don't know how to properly submit PRs for nixpkgs and besides, there's probably a dedicated maintainer for pam stuff (at least I imagine so).

Thanks!

**Comment by flokli** [comment]
From skimming through `nixos/modules/security/pam.nix`, I only see it once in `auth`, `password` and `session`.

There's a configuration similar to your proposed change at https://wiki.archlinux.org/index.php/pam_mount#Login_manager_configuration, skipping `pam_mount` if the service is `systemd-user` - I wonder why it only needs to be skipped there…

Could you elaborate a bit further, maybe with a minimal nixos configuration to reproduce the issue?
I could then follow-up, fix and add a NixOS test to it.

**Comment by flokli** [comment]
@voobscout can you provide some more info, as requested in https://github.com/NixOS/nixpkgs/issues/50447#issuecomment-462713434 ?

**Comment by benpye** [comment]
This is still an issue, and it is broader than just `pam_mount` unfortunately. I was playing with `pam_zfs_key` which even after this workaround doesn't successfully unmount (see: https://github.com/benpye/nixpkgs/commit/60a0554bc448fee11ced3c2a5b4e8d307f46a5f3). Ultimately the issue is that `pam_systemd` drops root, and then these workarounds all try and avoid issues by skipping the `systemd-user` session. This also doesn't really work because it means we unmount too early.

This is also tracked on the systemd side as https://github.com/systemd/systemd/issues/8598 and I opened an issue on the ZFS repo too https://github.com/openzfs/zfs/issues/12430

**Comment by NetaliDev** [comment]
I have fixed this a while ago in #182413 and #167215 so this should be solved now.

**Comment by Sk7Str1p3** [comment]
did it actually fix #21314 ?
